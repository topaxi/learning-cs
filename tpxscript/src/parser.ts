import { InputStream } from './input-stream'
import { TokenStream, Token } from './token-stream'

export const FALSE: Token<false> = { type: 'bool', value: false }
const PRECEDENCE: Record<string, number> = {
  '=': 2,
  '||': 3,
  '&&': 4,
  '<': 8,
  '>': 8,
  '>=': 8,
  '<=': 8,
  '==': 8,
  '!=': 8,
  '<<': 10,
  '>>': 10,
  '>>>': 10,
  '+': 11,
  '-': 11,
  '*': 20,
  '/': 20,
  '%': 20,
  '**': 25
}

export function parse(str: string) {
  let input = new InputStream(str)
  let tokens = new TokenStream(input)

  return parseTokens(tokens)
}

export function parseTokens(tokens: TokenStream) {
  let parser = new Parser(tokens)

  return parser.parseToplevel()
}

export class Parser {
  constructor(readonly tokens: TokenStream) {}

  isToken(type: string, ch?: string): Token | false {
    let token = this.tokens.peek()

    return (
      token !== null &&
      token.type === type &&
      (!ch || token.value === ch) &&
      token
    )
  }

  isPunctuation(ch?: string): Token | false {
    return this.isToken('punc', ch)
  }

  isKeyword(kw?: string): Token | false {
    return this.isToken('kw', kw)
  }

  isOperator(op?: string): Token | false {
    return this.isToken('op', op)
  }

  skipPunctuation(ch: string) {
    if (this.isPunctuation(ch) !== false) {
      this.tokens.next()
    } else {
      this.tokens.croak(`Expecting punctuation: "${ch}"`)
    }
  }

  skipKeyword(kw: string) {
    if (this.isKeyword(kw) !== false) {
      this.tokens.next()
    } else {
      this.tokens.croak(`Expecting keyword: "${kw}"`)
    }
  }

  skipOperator(op: string) {
    if (this.isOperator(op) !== false) {
      this.tokens.next()
    } else {
      this.tokens.croak(`Expecting operator: "${op}"`)
    }
  }

  unexpected(): never {
    return this.tokens.croak(
      `Unexpected token: ${JSON.stringify(this.tokens.peek())}`
    )
  }

  maybeBinary(left: any, myPrec: any): any {
    let token: any = this.isOperator()

    if (token !== false) {
      let hisPrec = PRECEDENCE[token.value as any]

      if (hisPrec > myPrec) {
        this.tokens.next()
        return this.maybeBinary(
          {
            type: token.value === '=' ? 'assign' : 'binary',
            operator: token.value,
            left,
            right: this.maybeBinary(this.parseAtom(), hisPrec)
          },
          myPrec
        )
      }
    }

    return left
  }

  delimited<T>(
    start: string,
    stop: string,
    separator: string,
    parser: (this: this) => T
  ): T[] {
    let a = []
    let first = true

    this.skipPunctuation(start)

    while (!this.tokens.eof()) {
      if (this.isPunctuation(stop) !== false) break
      if (first) {
        first = false
      } else {
        this.skipPunctuation(separator)
      }
      if (this.isPunctuation(stop) !== false) break
      a.push(parser.call(this))
    }
    this.skipPunctuation(stop)
    return a
  }

  parseCall(func: any) {
    return {
      type: 'call',
      func,
      args: this.delimited('(', ')', ',', this.parseExpression)
    }
  }

  parseVarname(): string {
    let name = this.tokens.next()

    if (!name || name.type !== 'var') {
      return this.tokens.croak('Expecting variable name')
    }

    return name.value
  }

  parseIf() {
    this.skipKeyword('if')

    let cond = this.parseExpression()

    if (this.isPunctuation('{') === false) this.skipKeyword('then')

    let then = this.parseExpression()
    let _else

    if (this.isKeyword('else') !== false) {
      this.tokens.next()
      _else = this.parseExpression()
    }

    return { type: 'if', cond, then, else: _else }
  }

  parseFn() {
    return {
      type: 'fn',
      name:
        this.tokens.peek()!.type === 'var'
          ? this.tokens.next()!.value
          : undefined,
      vars: this.delimited('(', ')', ',', this.parseVarname),
      body: this.parseExpression()
    }
  }

  parseBool() {
    return {
      type: 'bool',
      value: this.tokens.next()!.value === 'true'
    }
  }

  maybeCall(expr: () => Token) {
    let res = expr()

    return this.isPunctuation('(') !== false ? this.parseCall(res) : res
  }

  parseAtom() {
    return this.maybeCall(() => {
      if (this.isPunctuation('(') !== false) {
        this.tokens.next()
        let expr = this.parseExpression()
        this.skipPunctuation(')')
        return expr
      }

      if (this.isPunctuation('{') !== false) return this.parseProg()
      if (this.isKeyword('if') !== false) return this.parseIf()
      if (
        this.isKeyword('true') !== false ||
        this.isKeyword('false') !== false
      )
        return this.parseBool()
      if (this.isKeyword('fn') !== false) {
        this.tokens.next()
        return this.parseFn()
      }

      let token = this.tokens.next()!

      if (
        token.type === 'var' ||
        token.type === 'num' ||
        token.type === 'str'
      ) {
        return token
      }

      this.unexpected()
    })
  }

  parseToplevel(): { type: 'prog'; prog: Token[] } {
    let prog = []

    while (!this.tokens.eof()) {
      prog.push(this.parseExpression())

      if (!this.tokens.eof()) this.skipPunctuation(';')
    }

    return { type: 'prog', prog }
  }

  parseProg(): { type: 'prog'; prog: Token[] } | typeof FALSE {
    let prog = this.delimited('{', '}', ';', this.parseExpression)
    if (prog.length === 0) return FALSE
    if (prog.length === 1) return prog[0]
    return { type: 'prog', prog }
  }

  parseExpression(): any {
    return this.maybeCall(() => this.maybeBinary(this.parseAtom(), 0))
  }
}
