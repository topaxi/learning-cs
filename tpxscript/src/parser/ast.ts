export type NodeType =
  | 'num'
  | 'str'
  | 'bool'
  | 'var'
  | 'fn'
  | 'call'
  | 'if'
  | 'assign'
  | 'binary'
  | 'prog'
  | 'let'

export interface BaseNode {
  type: NodeType
}

export interface NumberNode extends BaseNode {
  type: 'num'
  value: number
}

export interface VarNode extends BaseNode {
  type: 'var'
  value: string
}

export type IdentifierNode = VarNode

export interface StringNode extends BaseNode {
  type: 'str'
  value: string
}

export interface BooleanNode extends BaseNode {
  type: 'bool'
  value: boolean
}

export interface CallNode extends BaseNode {
  type: 'call'
  func: Node
  args: any[]
}

export interface IfNode extends BaseNode {
  type: 'if'
  cond: Node
  then: Node
  else: Node | undefined
}

export interface FnNode extends BaseNode {
  type: 'fn'
  name: string | undefined
  vars: any[]
  body: Node
}

export interface BinaryNode extends BaseNode {
  type: 'binary' | 'assign'
  operator: string
  left: Node
  right: Node
}

export interface AssignNode extends BinaryNode {
  type: 'assign'
  operator: '='
}

export interface ProgNode extends BaseNode {
  type: 'prog'
  prog: Node[]
}

export interface LetNode extends BaseNode {
  type: 'let'
}

export type Node =
  | NumberNode
  | VarNode
  | StringNode
  | BooleanNode
  | CallNode
  | IfNode
  | FnNode
  | BinaryNode
  | AssignNode
  | ProgNode
  | LetNode
