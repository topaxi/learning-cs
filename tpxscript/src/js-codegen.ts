import { Token } from './token-stream'
import { FALSE } from './parser'

function assertNever(x: any): never {
  throw new Error(`Unexpected value ${x}`)
}

export class JsCodegen {
  generate(exp: Token): string {
    switch (exp.type) {
      case 'num':
      case 'str':
      case 'bool':
        return this.atom(exp)
      case 'binary':
        return this.binary(exp)
      case 'var':
        return this.var(exp)
      case 'assign':
        return this.assign(exp)
      case 'if':
        return this.if(exp)
      case 'fn':
        return this.fn(exp)
      case 'prog':
        return this.prog(exp)
      case 'call':
        return this.call(exp)
      default:
        return assertNever(exp.type)
    }
  }

  private atom(exp: Token): string {
    return JSON.stringify(exp.value)
  }

  private makeVar(name: string): string {
    return name
  }

  private var(exp: Token<string>): string {
    return this.makeVar(exp.value)
  }

  private binary(exp: Token | any): string {
    return `(${this.generate(exp.left)} ${exp.operator} ${this.generate(
      exp.right
    )})`
  }

  private assign(exp: Token): string {
    return this.binary(exp)
  }

  private fn(exp: Token | any): string {
    return `(function ${
      typeof exp.name === 'string' ? this.makeVar(exp.name) : ''
    }(${exp.vars.map(this.makeVar, this).join(', ')}) { return ${this.generate(
      exp.body
    )} })`
  }

  private prog(exp: Token | any): string {
    return `(${exp.prog.map(this.generate, this).join(',')})`
  }

  private call(exp: Token | any): string {
    return `${this.generate(exp.func)}(${exp.args
      .map(this.generate, this)
      .join(', ')})`
  }

  private if(exp: Token | any): string {
    return `(${this.generate(exp.cond)} ? ${this.generate(
      exp.then
    )} : ${this.generate(exp.else || FALSE)})`
  }
}
