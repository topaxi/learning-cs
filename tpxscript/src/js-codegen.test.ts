import * as prettier from 'prettier'
import { parse } from './parser'
import { JsCodegen } from './js-codegen'

function codegen(src: string) {
  return prettier.format(codegenUgly(src), {
    semi: false,
    parser: 'babel'
  })
}

function codegenUgly(src: string) {
  let gen = new JsCodegen()

  return gen.generate(parse(src) as any)
}

describe('tpxscript::JsCodegen', () => {
  test('should codegen numbers', () => {
    expect(codegenUgly('123.5')).toEqual('(123.5)')
  })

  test('should codegen strings', () => {
    expect(codegenUgly('"Hello, World!"')).toEqual('("Hello, World!")')
  })

  test('should codegen booleans', () => {
    expect(codegenUgly('true')).toEqual('(true)')
    expect(codegenUgly('false')).toEqual('(false)')
  })

  test('should codegen identifiers', () => {
    expect(codegenUgly('foo')).toEqual('(foo)')
  })

  test('should codegen function declarations', () => {
    expect(codegen('fn (x) x')).toMatchSnapshot()
    expect(codegen('fn (x) { x }')).toMatchSnapshot()
    expect(codegen('fn foo(x) x')).toMatchSnapshot()
    expect(codegen('fn foo(x) { x }')).toMatchSnapshot()
  })

  test('should codegen function calls', () => {
    expect(codegen('foo(x, 1)')).toMatchSnapshot()
  })

  test('should codegen conditionals', () => {
    expect(codegen('if foo then bar else baz')).toMatchSnapshot()
    expect(codegen('if foo then bar')).toMatchSnapshot()
    expect(codegen('if foo then if bar then baz')).toMatchSnapshot()
  })

  test('should codegen assignments', () => {
    expect(codegen('x = 10')).toMatchSnapshot()
  })

  test('should codegen arithmetic', () => {
    expect(codegenUgly('x + y')).toMatchSnapshot()
    expect(codegenUgly('x + y * z')).toMatchSnapshot()
  })

  test('should codegen logical operators', () => {
    expect(codegenUgly('x || y')).toEqual('((x || y))')
    expect(codegenUgly('x && y')).toEqual('((x && y))')
    expect(codegenUgly('x or y')).toEqual('((x || y))')
    expect(codegenUgly('x and y')).toEqual('((x && y))')
  })

  test('should codegen program', () => {
    let js = codegen(`
      fib = fn fib(n) if n < 3 then 1 else fib(n - 1) + fib(n - 2);
      ret = fib(10);
    `)
    expect(js).toMatchSnapshot()
    expect(new Function(`${js}; return ret`)()).toBe(55)
  })

  test.skip('should codegen blocks', () => {
    expect(
      codegen(`
      {
        a = 5;
        b = a * 2;
        a + b;
      };
    `)
    ).toMatchSnapshot()
  })

  test.skip('should codegen block variables', () => {
    expect(
      codegen(`
      let (a = 10, b = a * 10) {
        a + b
      }
    `)
    ).toMatchSnapshot()
  })
})
