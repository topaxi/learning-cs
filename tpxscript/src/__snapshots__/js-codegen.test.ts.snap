// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tpxscript::JsCodegen should codegen arithmetic 1`] = `"((x + (y * z)))"`;

exports[`tpxscript::JsCodegen should codegen assignments 1`] = `
"x = 10
"
`;

exports[`tpxscript::JsCodegen should codegen conditionals 1`] = `
"foo ? bar : baz
"
`;

exports[`tpxscript::JsCodegen should codegen conditionals 2`] = `
"foo ? bar : false
"
`;

exports[`tpxscript::JsCodegen should codegen function calls 1`] = `
"foo(x, 1)
"
`;

exports[`tpxscript::JsCodegen should codegen function declarations 1`] = `
";(function(x) {
  return x
})
"
`;

exports[`tpxscript::JsCodegen should codegen function declarations 2`] = `
";(function(x) {
  return x
})
"
`;

exports[`tpxscript::JsCodegen should codegen function declarations 3`] = `
";(function foo(x) {
  return x
})
"
`;

exports[`tpxscript::JsCodegen should codegen function declarations 4`] = `
";(function foo(x) {
  return x
})
"
`;
